src/core/apparmor-setup.c

```diff
@@ -0,0 +1,96 @@
+        _cleanup_(aa_kernel_interface_unrefp) aa_kernel_interface *interface = NULL;
+        const char *current_file;
+        const char *profile_file;
+        _cleanup_free_ char *current_profile = NULL;
+
+        if (!mac_apparmor_use()) {
+                log_debug("AppArmor either not supported by the kernel or disabled.");
+                return 0;
+        }
+
+        /* To enable LSM stacking a patch to the kernel is proposed to create a
+         * per-LSM subdirectory to distinguish between the LSMs. Therefore, we
+         * read the file from the LSM specific directory first and only if that
+         * fails the one from the generic directory.
+         */
+        FOREACH_STRING(current_file, "/proc/self/attr/apparmor/current", "/proc/self/attr/current") {
+                r = read_one_line_file(current_file, &current_profile);
+                if (r > 0)
+                        break;
+        }
+        if (current_profile == NULL) {
+                log_warning_errno(r, "Failed to get the current AppArmor profile of systemd from /proc/self/attr/current and /proc/self/attr/apparmor/current, ignoring: %m");
+                return 0;
+        }
+        if (!streq(current_profile, "unconfined")) {
+                log_debug("Systemd is already confined in an AppArmor profile.");
+                return 0;
+        }
+
+        FOREACH_STRING(profile_file, "/etc/systemd/profile.apparmor", "/usr/lib/systemd/profile.apparmor") {
+                fd = open(profile_file, O_RDONLY | O_CLOEXEC);
+                if (fd < 0) {
+                        if (errno == ENOENT)
+                                continue;
+                        log_error_errno(errno, "Failed to open binary AppArmor profile file %s, ignoring: %m", profile_file);
+                } else
+                        break;
+        }
+        /* Do not log anything if both files do not exist, because we assume the
+         * user does not want to use AppArmor for systemd itself in this case.
+         * Other errors have already been logged above.
+         */
+        if (fd < 0 ) {
+                return 0;
+        }
+
+        r = aa_kernel_interface_new(&interface, NULL, NULL);
+        if (r < 0) {
+                log_warning_errno(errno, "Failed to get a new AppArmor kernel interface: %m");
+                return 0;
+        }
+
+        r = aa_kernel_interface_load_policy_from_fd(interface, fd);
+        if (r < 0) {
+                log_warning_errno(errno, "Failed to load AppArmor profile from file %s into the kernel: %m", profile_file);
+                return 0;
+        } else
+                log_info("AppArmor profile file %s loaded.", profile_file);
+
+        r = aa_change_profile("systemd");
+        if (r < 0) {
+                if (errno == ENOENT)
+                        log_error("Failed to change to AppArmor profile systemd. Please ensure that the profile file %s contains a profile with that name.", profile_file);
```

Comment:
doesn't matter much, but i'd really use `log_error_errno(errno, …` here too. When we do structured logging to the journal then the error passed like that is added as a structured field `ERRNO`, hence whenever we have an error we should use the log_xyz_errno() flavour of calls, to pass it along, even if the format string doesn't include `%m`.